File: engine/governance/GovernanceKernel.js
/**
 * Governance Kernel — CKL + contracts → Decision
 */

import { resolveAuthority } from "../constitution/contracts.js";
import { CHARTER } from "../constitution/charter.js";
import {
  ConstitutionalKnowledgeLayer,
  resolveDecision,
} from "./ConstitutionalKnowledgeLayer.js";

export class GovernanceKernel {
  /**
   * @param {object} options
   * @param {import('./ConstitutionalKnowledgeLayer.js').ConstitutionalKnowledgeLayer} options.ckl
   */
  constructor({ ckl } = {}) {
    this.ckl = ckl ?? new ConstitutionalKnowledgeLayer([]);
    this.charterId = CHARTER.id;
  }

  /**
   * evaluateIntent(intent, evidence) → Decision
   */
  evaluateIntent(intent, evidence) {
    const worldId = intent?.world ?? intent?.constraints?.worldId ?? "*";
    const policies = this.ckl.GetPoliciesForWorld(worldId);
    const precedents = this.ckl.GetPrecedents(intent);
    const decision = resolveDecision(intent, evidence, policies, precedents);

    const enriched = {
      ...decision,
      charterId: this.charterId,
      intentId: intent?.id ?? null,
      worldId,
      policiesApplied: policies.policies.map((p) => p.id),
      precedentCount: precedents.length,
    };

    this.ckl.recordPrecedent({ intent, decision: enriched });
    return enriched;
  }

  resolveAuthority(actor, action) {
    return resolveAuthority(actor, action);
  }
}

File: unity/GovernedUnityProject/Assets/Engine/Shared/Governance/GovernanceKernel.cs
// mirrors engine/governance/GovernanceKernel.cs
using System.Linq;
using SovereignX.CIEMS.Engine.Runtime;

namespace SovereignX.CIEMS.Engine.Governance
{
    public sealed class GovernanceKernel
    {
        readonly IConstitutionalKnowledgeLayer _ckl;
        public string CharterId { get; set; } = "charter.4dce.v1";

        public GovernanceKernel(IConstitutionalKnowledgeLayer ckl)
        {
            _ckl = ckl;
        }

        public Decision EvaluateIntent(IntentRecord intent, EvidenceBundle evidence)
        {
            var worldId = intent?.World ?? "*";
            if (intent?.Constraints != null &&
                intent.Constraints.TryGetValue("worldId", out var w) &&
                w is string ws &&
                !string.IsNullOrEmpty(ws))
            {
                worldId = intent.World ?? ws;
            }

            var policies = _ckl.GetPoliciesForWorld(worldId);
            var precedents = _ckl.GetPrecedents(intent);
            var decision = DecisionEngine.Resolve(intent, evidence, policies, precedents);

            decision.CharterId = CharterId;
            decision.IntentId = intent?.Id;
            decision.WorldId = worldId;
            decision.PoliciesApplied = policies.Policies.Select(p => p.Id).ToList();
            decision.PrecedentCount = precedents.Count;

            _ckl.RecordPrecedent(intent, decision);
            return decision;
        }
    }
}
